### PR TITLE
[Woo POS - payments onboarding] Show onboarding with POS modal instead of sheet

### LINKED ISSUE #14209: [Woo POS - payments onboarding] Use `posModal` instead of `sheet` for onboarding modal
More context and info from Josh on how to make the onboarding view compatible with `posModal`: https://github.com/woocommerce/woocommerce-ios/pull/14185#issuecomment-2434784549

### PR DESCRIPTION
**Pull request by jaclync**
Closes: #14209 

## Why

Previously, the onboarding UI is shown in the default sheet in SwiftUI which doesn't match the style of other modals in POS. For example, the modal about exiting POS, and the modal about the product type. This PR updates the presentation to show the same onboarding view in `posModal` instead of `sheet`.

## How

A new SwiftUI view `PointOfSaleCardPresentPaymentOnboardingView` was created as a POS-specific view that contains a HStack with a X CTA to dismiss the modal, and the same onboarding view as before. This way, we can customize the modal content and potentially swap out the onboarding UI for any future potential customizations. Onboarding view model's `showURL` can also be handled in the new view's view model, removing the need for a URL property in `TotalsViewModel`.

For `CardPresentPaymentsOnboardingViewModel` to be used with `posModal`, `Equatable` conformance is required and implemented in https://github.com/woocommerce/woocommerce-ios/pull/14224/commits/61392d376c0530a2810df4d5614b0298b41d004d.

## Steps to reproduce

Prerequisite: store is eligible for POS, and onboarding isn't completed (missing some requirements, or plugin not activated with a Shop Manager role). Alternatively, you can hard code the state by returning `.stripeAccountOverdueRequirement(plugin: .wcPay)` in `CardPresentPaymentsOnboardingUseCase.checkOnboardingState`.

- Switch to store/user in the prerequisite, if needed
- Go to Menu > POS Mode
- Tap `Connect your reader` --> an onboarding modal should be shown shortly with an X CTA at the top right, with a `Learn more` CTA and `Contact us` CTA
- Tap outside of the modal --> the modal should be dismissed
- Tap `Connect your reader` again
- Tap `Learn more` --> a Safari sheet should be shown about the onboarding error
- Dismiss the webview
- Tap `Contact us` --> the onboarding modal should be dismissed, then support form is presented
- Tap `Connect your reader` again
- Repeat the steps above to tap on the 2 CTAs
- Optionally & if not hard-coding the error state, try resolving the onboarding error(s) to proceed to reader connection

## Testing information

I tested in iPad Pro 11in iOS 17.5 simulator and 17.5.1 device with Xcode 16, in several stores in different states.

## Screenshots

https://github.com/user-attachments/assets/b4cdb944-0fee-4e23-a6da-0e30a7f5a53e

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person Payments/CardPresentPaymentsOnboardingViewModel.swift
Comment (iamgabrielma) [inline WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person Payments/CardPresentPaymentsOnboardingViewModel.swift]: ❓ A bit of a general question, since we cannot really fully conform the protocol and the closures, would still be worth to approximate it? For example:

```suggestion
extension CardPresentPaymentsOnboardingViewModel: Equatable {
    static func == (lhs: CardPresentPaymentsOnboardingViewModel, rhs: CardPresentPaymentsOnboardingViewModel) -> Bool {
        lhs.state == rhs.state &&
        lhs.userIsAdministrator == rhs.userIsAdministrator &&
        lhs.learnMoreURL == rhs.learnMoreURL
        lhs.useCase.state == rhs.useCase.state &&
        (lhs.showSupport != nil) == (rhs.showSupport != nil) &&
        (lhs.showURL != nil) == (rhs.showURL != nil)
    }
}
```
I don't have any specific use case in mind where performing an equal check between two onboarding view models could pass as false positive due to missing properties, it's more of "I'm curious" 😅

File: WooCommerce/WooCommerceTests/POS/ViewModels/PointOfSaleCardPresentPaymentOnboardingViewModelTests.swift
Comment (iamgabrielma) [inline WooCommerce/WooCommerceTests/POS/ViewModels/PointOfSaleCardPresentPaymentOnboardingViewModelTests.swift]: I think we don't need this one:
```suggestion
```

File: WooCommerce/WooCommerceTests/POS/ViewModels/PointOfSaleCardPresentPaymentOnboardingViewModelTests.swift
Comment (iamgabrielma) [inline WooCommerce/WooCommerceTests/POS/ViewModels/PointOfSaleCardPresentPaymentOnboardingViewModelTests.swift:23]: Should we add an assertion here just to assure initial state? 
```suggestion
        let sut = PointOfSaleCardPresentPaymentOnboardingViewModel(onboardingViewModel: onboardingViewModel, onDismissTap: nil)
        XCTAssertNil(sut.onboardingURL)
```

File: WooCommerce/Classes/POS/Presentation/Payments Onboarding/PointOfSaleCardPresentPaymentOnboardingViewModel.swift
Comment (iamgabrielma) [inline WooCommerce/Classes/POS/Presentation/Payments Onboarding/PointOfSaleCardPresentPaymentOnboardingViewModel.swift:4]: ❓ Question: Would it be the same to have internal `let` that `private(set) var` here? I'm assuming is exactly the same because we're not modifying it after initialization?

File: WooCommerce/Classes/POS/Presentation/Payments Onboarding/PointOfSaleCardPresentPaymentOnboardingView.swift
Comment (iamgabrielma) [inline WooCommerce/Classes/POS/Presentation/Payments Onboarding/PointOfSaleCardPresentPaymentOnboardingView.swift]: 💯 so easy to forget about accessibility!

File: WooCommerce/Classes/POS/Presentation/Payments Onboarding/PointOfSaleCardPresentPaymentOnboardingView.swift
Comment (iamgabrielma) [inline WooCommerce/Classes/POS/Presentation/Payments Onboarding/PointOfSaleCardPresentPaymentOnboardingView.swift:15]: ❓  Question: Would it be worth to use `WebViewSheet` rather than `SafariSheetView` here? I haven't checked further but I think we can leverage the authenticated web view with the former.

**Review by iamgabrielma — APPROVED** [review]
Looks good and works well :shipit: 

✅ Modal shown
✅ Dismissed by drag
✅ Dismissed by tap outside
✅ Learn more works as intended
✅ Contact support works well and dismisses the modal automatically 
✅ Connection flow successful

File: WooCommerce/Classes/POS/Presentation/Payments Onboarding/PointOfSaleCardPresentPaymentOnboardingView.swift
Comment (joshheald) [inline WooCommerce/Classes/POS/Presentation/Payments Onboarding/PointOfSaleCardPresentPaymentOnboardingView.swift]: Consider using the `.posModalCloseButton(action: accessibilityLabel:)` modifier instead – you can just put it on the Onboarding view at the end somewhere and it saves repeating this. See `PointOfSaleCardPresentPaymentScanningForReadersView` for an example

File: WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person Payments/CardPresentPaymentsOnboardingViewModel.swift
Comment (joshheald) [inline WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person Payments/CardPresentPaymentsOnboardingViewModel.swift]:
```suggestion
        lhs.learnMoreURL == rhs.learnMoreURL &&
        lhs.shouldShow == rhs.shouldShow
```

File: WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person Payments/CardPresentPaymentsOnboardingViewModel.swift
Comment (joshheald) [reply WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person Payments/CardPresentPaymentsOnboardingViewModel.swift]: This might be worth doing, and would slightly improve the conformance.

A bigger risk is sometimes that other properties could be added in future and missed from the equatable conformance, because the person adding them doesn't realise there's a custom function for it.

One thing I've added recently is a test function for counting properties – if you add a test asserting that the property count matches what it is right now, with a message saying "Please update the equatable conformance if this test fails", you give a chance to catch that issue in future.

[See `PointOfSaleCardPresentPaymentConnectingFailedUpdateAddressAlertViewModelTests` for an example.](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/WooCommerceTests/POS/Card%20Present%20Payments/Connection%20Alerts/PointOfSaleCardPresentPaymentConnectingFailedUpdateAddressAlertViewModelTests.swift)

File: WooCommerce/Classes/POS/Presentation/PointOfSaleDashboardView.swift
Comment (joshheald) [inline WooCommerce/Classes/POS/Presentation/PointOfSaleDashboardView.swift]: I'm confused by this... in the description it says we should be able to tap outside the modal to dismiss it, but this modifier prevents that happening.

Payment modals can be dismissed by tapping outside, which was a design request, so I think onboarding ones should work the same way.

You could extract the frame code from `PointOfSaleCardPresentPaymentAlert` to share it between payment alerts and onboarding – they should probably have the same size frames for visual consistency. That might be something to do once we redesign them though, the old style may not fit because it's so wordy.

```suggestion
```

Comment (jaclync) [reply WooCommerce/Classes/POS/Presentation/PointOfSaleDashboardView.swift]: > in the description it says we should be able to tap outside the modal to dismiss it, but this modifier prevents that happening.

Thanks for catching this one, I originally meant "tapping outside the modal should **not** dismiss it" so it's very confusing 😓 

> Payment modals can be dismissed by tapping outside, which was a design request, so I think onboarding ones should work the same way.

I see, yea it makes sense to make it possible to dismiss the onboarding by tapping outside if there was a design request. For more context, the interactive dismissal was first disabled with the feedback in https://github.com/woocommerce/woocommerce-ios/pull/14185#discussion_r1812064752 to prevent the onboarding from being dismissed accidentally. But from more testing, the webview that shows the WooPayments/Stripe form can still be dismissed interactively and some changes are required within the onboarding view which affects the pre-existing onboarding flows.

File: WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person Payments/CardPresentPaymentsOnboardingViewModel.swift
Comment (jaclync) [reply WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person Payments/CardPresentPaymentsOnboardingViewModel.swift]: Great catch, I didn't realize there are variables below the functions 😅 was it intentional to organize the functions & variables this way in payments at least?

Updated in https://github.com/woocommerce/woocommerce-ios/pull/14224/commits/84455c4377b0ac4fc47d72adbdfc723bfabc9418.

File: WooCommerce/WooCommerceTests/POS/ViewModels/PointOfSaleCardPresentPaymentOnboardingViewModelTests.swift
Comment (jaclync) [reply WooCommerce/WooCommerceTests/POS/ViewModels/PointOfSaleCardPresentPaymentOnboardingViewModelTests.swift]: Nice catch, removed in https://github.com/woocommerce/woocommerce-ios/pull/14224/commits/4ab6e1aad87a93492315e0bc8e8ac8c6923a5b65.

File: WooCommerce/WooCommerceTests/POS/ViewModels/PointOfSaleCardPresentPaymentOnboardingViewModelTests.swift
Comment (jaclync) [reply WooCommerce/WooCommerceTests/POS/ViewModels/PointOfSaleCardPresentPaymentOnboardingViewModelTests.swift:23]: Yea makes sense! Added in https://github.com/woocommerce/woocommerce-ios/pull/14224/commits/4ab6e1aad87a93492315e0bc8e8ac8c6923a5b65.

File: WooCommerce/Classes/POS/Presentation/Payments Onboarding/PointOfSaleCardPresentPaymentOnboardingViewModel.swift
Comment (jaclync) [reply WooCommerce/Classes/POS/Presentation/Payments Onboarding/PointOfSaleCardPresentPaymentOnboardingViewModel.swift:4]: `private(set) var` still allows the class to change its value internally, while `let` disallows it. Since the onboarding view model `CardPresentPaymentsOnboardingViewModel ` source of truth comes from above (TotalsVM for now), it's not expected to be mutated internally.

File: WooCommerce/Classes/POS/Presentation/Payments Onboarding/PointOfSaleCardPresentPaymentOnboardingView.swift
Comment (jaclync) [reply WooCommerce/Classes/POS/Presentation/Payments Onboarding/PointOfSaleCardPresentPaymentOnboardingView.swift:15]: Good point - I followed the implementation in Menu > Payments to show the URL in a Safari sheet

https://github.com/woocommerce/woocommerce-ios/blob/5d92a93eaf73efa69b97dcf8fca3b15ac128570d/WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person%20Payments/Payments%20Menu/InPersonPaymentsMenuViewModel.swift#L238-L241

But I will do more testing when logging in with WPCOM and site credentials to see if it's possible to skip the login step when completing the WooPayments/Stripe form in the webview. Added a subtask: https://github.com/woocommerce/woocommerce-ios/issues/14233

File: WooCommerce/Classes/POS/Presentation/Payments Onboarding/PointOfSaleCardPresentPaymentOnboardingViewModel.swift
Comment (iamgabrielma) [reply WooCommerce/Classes/POS/Presentation/Payments Onboarding/PointOfSaleCardPresentPaymentOnboardingViewModel.swift:4]: That makes sense, thanks for the clarification

File: WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person Payments/CardPresentPaymentsOnboardingViewModel.swift
Comment (jaclync) [reply WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person Payments/CardPresentPaymentsOnboardingViewModel.swift]: I'm more used to comparing what represents the identity of a class to trigger a redraw - like the state in this case - without the closures and internal dependencies, but it looks safer to include the closure nullability checks. Great suggestion on adding a test case to fail when new properties are added. Updates in https://github.com/woocommerce/woocommerce-ios/pull/14224/commits/66dca3d68202c6773436835f7066559ad20fb799.

File: WooCommerce/Classes/POS/Presentation/PointOfSaleDashboardView.swift
Comment (jaclync) [reply WooCommerce/Classes/POS/Presentation/PointOfSaleDashboardView.swift]: Interactive dismissal enabled in https://github.com/woocommerce/woocommerce-ios/pull/14224/commits/9da408ad5227ae23f8c664de6f7a64314948c80f.

> You could extract the frame code from PointOfSaleCardPresentPaymentAlert to share it between payment alerts and onboarding – they should probably have the same size frames for visual consistency. That might be something to do once we redesign them though, the old style may not fit because it's so wordy.

Good to know, added in https://github.com/woocommerce/woocommerce-ios/pull/14224/commits/107699e9bae23118124117a6f22bfe99de891d0b for POS onboarding and the replacement for POS payment modals is in https://github.com/woocommerce/woocommerce-ios/pull/14224/commits/1b7dcfc847421fa54d6aceb542b6db3eab0501a9.

The payment alert UI stays the same:

File: WooCommerce/Classes/POS/Presentation/Payments Onboarding/PointOfSaleCardPresentPaymentOnboardingView.swift
Comment (jaclync) [reply WooCommerce/Classes/POS/Presentation/Payments Onboarding/PointOfSaleCardPresentPaymentOnboardingView.swift]: Nice, updated in https://github.com/woocommerce/woocommerce-ios/pull/14224/commits/107699e9bae23118124117a6f22bfe99de891d0b.

**Comment by jaclync** [comment]
Thanks for the suggestions/questions @iamgabrielma @joshheald that I wasn't aware of before, I believe I've responded to all the comments with code changes and updated screencast in the PR description. Ready for another pass!

**Comment by jaclync** [comment]
@iamgabrielma @joshheald 👋 would one of you like to review again? I'll wait for feedback and merge it on Friday if I don't see further comments (onboarding is behind a local feature flag), but feel free to leave comments after merging and I'll address them separately.

**Comment by iamgabrielma** [comment]
> I'll wait for feedback and merge it on Friday if I don't see further comments (onboarding is behind a local feature flag), but feel free to leave comments after merging and I'll address them separately.

LGTM 🚢

**Pull request merged**
